drivers/network/netio/netio.c

```diff
@@ -0,0 +1,903 @@
+    void *user_context;         /* parameter for callbacks, opaque */
+    struct _UNICODE_STRING TdiName;     /* \\Devices\\Tcp, \\Devices\\Udp */
+
+    struct sockaddr LocalAddress;
+    PFILE_OBJECT LocalAddressFile;
+    HANDLE LocalAddressHandle;
+
+    struct sockaddr RemoteAddress;      /* Remeber latest remote connection */
+
+    /* Those exist for connection oriented (TCP/IP) sockets only */
+    PFILE_OBJECT ConnectionFile;        /* Returned by TdiOpenConnectionEndpointFile() */
+    HANDLE ConnectionHandle;    /* Returned by TdiOpenConnectionEndpointFile() */
+
+    /* Incoming connection callback function: */
+    const struct _WSK_CLIENT_LISTEN_DISPATCH *ListenDispatch;
+    int CallbackMask;
+
+    int Flags;                  /* SO_REUSEADDR, ... see ws2def.h */
+    /* TODO: we have refcounting now, this should go away: */
+    LIST_ENTRY PendingUserIrps; /* Irps we got from our user.
+                                   For cancelling in WskClose() */
+    int RefCount;               /* See SocketGet/SocketPut TODO: this should be atomic */
+};
+
+struct NetioContext
+{
+    PIRP UserIrp;
+    struct _WSK_SOCKET_INTERNAL *socket;
+};
+
+void
+SocketGet(struct _WSK_SOCKET_INTERNAL *s)
+{
+    s->RefCount++;
+}
+
+void
+SocketPut(struct _WSK_SOCKET_INTERNAL *s)
+{
+    /* TODO: get some spinlock */
+    s->RefCount--;
+    if (s->RefCount == 0)
+    {
+        if (s->LocalAddressHandle != INVALID_HANDLE_VALUE)
+        {
+            ZwClose(s->LocalAddressHandle);
+            s->LocalAddressHandle = INVALID_HANDLE_VALUE;
+            s->LocalAddressFile = NULL;
+        }
+        if (s->ConnectionHandle != INVALID_HANDLE_VALUE)
+        {
+            ZwClose(s->ConnectionHandle);
+            s->ConnectionHandle = INVALID_HANDLE_VALUE;
+            s->ConnectionFile = NULL;
+        }
+        /* TODO: Free TdiName and others */
+        ExFreePoolWithTag(s, 'SOCK');
+    }
+}
+
+/* This just sets the status to pending. It is been called
+ * by the IoCallDriver() function via the MajorFunction
+ * dispatch table in the DriverObject. We need this because
+ * we want to use IoCallDriver to consume one stack location
+ * of the irp. If we do not do so the completion routine of
+ * the upper driver never gets called.
+ */
+
+static NTSTATUS NTAPI
+DummyHandler(PDEVICE_OBJECT Device, PIRP Irp)
+{
+    Irp->IoStatus.Status = STATUS_PENDING;
+
+    return STATUS_PENDING;
+}
+
+struct DummyDeviceExtension
+{
+    int x;
+};
+
+PDEVICE_OBJECT DummyDeviceObject;
+
+static NTSTATUS
+DummyCallDriver(PIRP Irp)
+{
+    return IoCallDriver(DummyDeviceObject, Irp);
+}
+
+NTSTATUS NTAPI
+DriverEntry(PDRIVER_OBJECT DriverObject, PUNICODE_STRING RegistryPath)
+{
+    int i;
+    NTSTATUS status;
+    UNICODE_STRING nameUnicode;
+
+    DbgPrint("netio.sys DriverEntry compiled " __DATE__ " " __TIME__ "\n");
+
+    for(i = 0; i <= IRP_MJ_MAXIMUM_FUNCTION; i++)
+        DriverObject->MajorFunction[i] = DummyHandler;
```

Comment:
Ok. I have to admit that even after 7 years of Windows driver development I still don't understand all aspects of IRPs, driver objects and so on.

I now removed the DummyXXX mechanism because calling IoSetNextIrpStackLocation() also gives the desired result (that the I/O completion of the driver that calls us gets called). So the DriverObject is also deleted from the NETIO driver.

Reply:
Hi this piece of code was removed, resolving conversation.